packages/dcos-integration-test/extra/test_dcos_diagnostics.py

```diff
@@ -640,35 +536,139 @@ def _get_dcos_diagnostics_health(z: zipfile.ZipFile, item: str):
+                "Job is not progressing"
+            )
+    last_datapoint['value'] = percent_done
+    last_datapoint['time'] = datetime.datetime.now()
+
+    return not job_running
+
+
+# sometimes it may take extra few seconds to list bundles after the job is finished.
+@retrying.retry(stop_max_delay=5000)
+def wait_for_diagnostics_list(dcos_api_session):
+    assert _get_bundle_list(dcos_api_session), 'get a list of bundles timeout'
+
+
+def validate_node(nodes):
+    assert isinstance(nodes, list), 'input argument must be a list'
+    assert len(nodes) > 0, 'input argument cannot be empty'
+    required_fields = ['host_ip', 'health', 'role']
+
+    for node in nodes:
+        assert len(node) == len(required_fields), 'node should have the following fields: {}. Actual: {}'.format(
+            ', '.join(required_fields), node)
+        for required_field in required_fields:
+            assert required_field in node, '{} must be in node. Actual: {}'.format(required_field, node)
+
+        # host_ip, health, role fields cannot be empty
+        assert node['health'] in [0, 1], 'health must be 0 or 1'
+        assert node['host_ip'], 'host_ip cannot be empty'
+        assert node['role'], 'role cannot be empty'
+
+
+def validate_units(units):
+    assert isinstance(units, list), 'input argument must be list'
+    assert len(units) > 0, 'input argument cannot be empty'
+    required_fields = ['id', 'name', 'health', 'description']
+
+    for unit in units:
+        assert len(unit) == len(required_fields), 'a unit must have the following fields: {}. Actual: {}'.format(
+            ', '.join(required_fields), unit)
+        for required_field in required_fields:
+            assert required_field in unit, 'unit response must have field: {}. Actual: {}'.format(required_field, unit)
+
+        # a unit must have all 3 fields not empty
+        assert unit['id'], 'id field cannot be empty'
+        assert unit['name'], 'name field cannot be empty'
+        assert unit['health'] in [0, 1], 'health must be 0 or 1'
+        assert unit['description'], 'description field cannot be empty'
+
+
+def validate_unit(unit):
+    assert isinstance(unit, dict), 'input argument must be a dict'
+
+    required_fields = ['id', 'health', 'output', 'description', 'help', 'name']
+    assert len(unit) == len(required_fields), 'unit must have the following fields: {}. Actual: {}'.format(
+        ', '.join(required_fields), unit)
+    for required_field in required_fields:
+        assert required_field in unit, '{} must be in a unit. Actual: {}'.format(required_field, unit)
+
+    # id, name, health, description, help should not be empty
+    assert unit['id'], 'id field cannot be empty'
+    assert unit['name'], 'name field cannot be empty'
+    assert unit['health'] in [0, 1], 'health must be 0 or 1'
+    assert unit['description'], 'description field cannot be empty'
+    assert unit['help'], 'help field cannot be empty'
```

Comment:
Future improvement for another PR: `validate_unit()` and `validate_units()` check a lot of the same stuff. Either `validate_units()` should call `validate_unit()` inside its for loop, or we should get rid of one and use the other in its place. Looks like `validate_unit()` only has one call site.

Reply:
I did it but `validate_units` could not use `validat_unit`. `validate_units` has less fields than `unit`. I'm not sure if we want to loose our assertions just to have clean tests.

Reply:
Oh interesting. It's unclear to me why one would test more fields than the other.

Reply:
One endpoint returns more data. No idea why